File: C#-OOP-Basics/Problems Course Instance 2018/Polymoprhism/Vehicles/Models/Vehicle.cs
public abstract class Vehicle
{
    private double fuel;
    private double distanceTravelled;

    protected Vehicle(double fuel, double fuelConsumption)
    {
        this.Fuel = fuel;
        this.FuelConsumptiomPerKm = fuelConsumption;
    }

    public double Fuel
    {
        get
        {
            return this.fuel;
        }
        set
        {
            this.fuel = value;
        }
    }

    public double DistanceTravelled
    {
        get
        {
            return this.distanceTravelled;
        }
        private set
        {
            this.distanceTravelled = value;
        }
    }

    public double FuelConsumptiomPerKm { get; protected set; }

    public virtual void Refuel(double fuelQuantity)
    {
        this.Fuel += fuelQuantity;
    }

    public virtual string Drive(double Km)
    {
        double fuelNeeded = Km * this.FuelConsumptiomPerKm;

        if (Fuel - fuelNeeded >= 0)
        {
            this.DistanceTravelled += Km;
            this.Fuel -= fuelNeeded;

            return  $"{this.GetType().Name} travelled {Km} km";
        }
        else
        {
            return $"{this.GetType().Name} needs refueling";
        }
    }
}

File: Java-OOP-Basics/Problems Course Instance 2016_2017/src/Lesson04Polymorphism/Exercise/pr01_vehicles/Vehicle.java
package Lesson04Polymorphism.Exercise.pr01_vehicles;

public abstract class Vehicle {

    private double fuelQuantity;
    private double fuelConsumptionPerKm;

    protected Vehicle(double fuelQuantity, double fuelConsumptionPerKm) {
        this.setFuelQuantity(fuelQuantity);
        this.setFuelConsumptionPerKm(fuelConsumptionPerKm);
    }

    protected final double getFuelQuantity() {
        return this.fuelQuantity;
    }

    protected void setFuelQuantity(double fuelQuantity) {
        this.fuelQuantity = fuelQuantity;
    }

    protected double getFuelConsumptionPerKm() {
        return this.fuelConsumptionPerKm;
    }

    private void setFuelConsumptionPerKm(double fuelConsumptionPerKm) {
        this.fuelConsumptionPerKm = fuelConsumptionPerKm;
    }

    public abstract void refuel(double fuel);

    public abstract void drive(double distance);
}
